test-app/tests/integration/environment-test.js

```diff
@@ -0,0 +1,364 @@
+
+  exclaimTest('array mutation', {
+    env: { foo: ['bar', 'baz'] },
+    ui: {
+      bound: { $bind: 'foo' },
+      async $validate(assert) {
+        set(this.env, 'foo.1', 'oops');
+        assert.strictEqual(this.config.bound[1], 'oops');
+        assert.strictEqual(get(this.env, 'foo.1'), 'oops');
+        set(this.config.bound, '0', 'oops again');
+        assert.strictEqual(get(this.env, 'foo.0'), 'oops again');
+      },
+    },
+  });
+
+  exclaimTest('HTML-safe strings', {
+    env: { foo: htmlSafe('hello') },
+    ui: {
+      async $validate(assert) {
+        assert.deepEqual(get(this.env, 'foo'), htmlSafe('hello'));
+      },
+    },
+  });
+
+  exclaimTest('simple binding resolution', {
+    env: { foo: 'bar' },
+    ui: {
+      baz: { $bind: 'foo' },
+      async $validate(assert) {
+        assert.strictEqual(get(this.config, 'baz'), 'bar');
+
+        set(this.env, 'foo', 'qux');
+        assert.strictEqual(get(this.config, 'baz'), 'qux');
+
+        set(this.config, 'baz', 'fizz');
+        assert.strictEqual(get(this.env, 'foo'), 'fizz');
+      },
+    },
+  });
+
+  exclaimTest('complex binding resolution', {
+    env: { x: { y: 'foo', z: 'bar' } },
+    resolveFieldMeta: (path) => path,
+    ui: {
+      other: { $bind: 'x' },
+      stillMore: { $bind: 'x.z' },
+      $validate(assert) {
+        assert.strictEqual(this.config.stillMore, 'bar');
+        assert.strictEqual(this.env.metaForField('x.y'), 'x.y');
+        assert.strictEqual(this.env.metaForField(this.config, 'other'), 'x');
+        assert.strictEqual(
+          this.env.metaForField(this.config, 'stillMore'),
+          'x.z'
+        );
+
+        set(this.env, 'x.z', 123);
+        assert.strictEqual(this.config.other.z, 123);
+        assert.strictEqual(this.config.stillMore, 123);
+
+        set(this.config, 'other.z', 234);
+        assert.strictEqual(this.config.other.z, 234);
+        assert.strictEqual(this.config.stillMore, 234);
+
+        set(this.config, 'stillMore', 345);
+        assert.strictEqual(this.config.other.z, 345);
+        assert.strictEqual(this.config.stillMore, 345);
+      },
+    },
+  });
+
+  exclaimTest('value reference resolution', {
+    env: { foo: 'bar' },
+    resolveFieldMeta: (path) => path,
+    ui: {
+      value: { key: { $bind: 'foo' } },
+      async $validate(assert) {
+        const value = this.config.value;
+        assert.strictEqual(get(value, 'key'), 'bar');
+        assert.strictEqual(this.env.metaForField(value, 'key'), 'foo');
+
+        set(value, 'key', 'ok');
+        assert.strictEqual(get(this.env, 'foo'), 'ok');
+      },
+    },
+  });
+
+  exclaimTest('chained reference resolution', {
+    env: { foo: 'bar' },
+    resolveFieldMeta: (path) => path,
+    ui: {
+      value: { key: { child: { $bind: 'foo' } } },
+      async $validate(assert) {
+        const value = this.config.value;
+        const subvalue = get(value, 'key');
+        assert.strictEqual(get(subvalue, 'child'), 'bar');
```

Comment:
Are these `.get`s still necessary for subvalues? Same question for the "array values" test below.